pennylane/tape/interfaces/jax.py

```diff
@@ -0,0 +1,116 @@
+
+    Alternatively, the jax interface can be dynamically applied to existing
+    quantum tapes via the :meth:`~.apply` class method. This modifies the
+    tape **in place**.
+
+    Once created, the autograd interface can be used to perform quantum-classical
+    differentiable programming.
+
+    .. note::
+
+        If using a device that supports native autograd computation and backpropagation, such as
+        :class:`~.DefaultQubitJAX`, the Autograd interface **does not need to be applied**. It
+        is only applied to tapes executed on non-JAX compatible devices.
+
+    **Example**
+
+    Once a JAX quantum tape has been created, it can be differentiated using JAX:
+
+    .. code-block:: python
+
+        tape = JAXInterface.apply(JacobianTape())
+
+        with tape:
+            qml.Rot(0, 0, 0, wires=0)
+            expval(qml.PauliX(0))
+
+        def cost_fn(x, y, z, device):
+            tape.set_parameters([x, y ** 2, y * np.sin(z)], trainable_only=False)
+            return tape.execute(device=device)
+
+    >>> x = jnp.array(0.1, requires_grad=False)
+    >>> y = jnp.array(0.2, requires_grad=True)
+    >>> z = jnp.array(0.3, requires_grad=True)
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> cost_fn(x, y, z, device=dev)
+    [0.03991951]
+    >>> jac_fn = jax.vjp(cost_fn)
+    >>> jac_fn(x, y, z, device=dev)
+    DeviceArray[[ 0.39828408, -0.00045133]]
+    """
+
+    # pylint: disable=attribute-defined-outside-init
+    dtype = jnp.float64
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "jax"
+
+    def _execute(self, params, device):
+        # TODO(chase): Add support for this.
+        if len(self.observables) != 1:
+            raise ValueError("Only one return type is supported currently")
+        return_type = self.observables[0].return_type
+        if return_type is not Variance and return_type is not Expectation:
+            raise ValueError(
+                f"Only Variance and Expectation returns are support, given {return_type}"
+            )
+        exec_fn = partial(self.execute_device, device=device)
+
+        return host_callback.call(
+            exec_fn, params, result_shape=jax.ShapeDtypeStruct((1,), JAXInterface.dtype)
+        )
+
+    @classmethod
+    def apply(cls, tape):
+        """Apply the JAX interface to an existing tape in-place.
+
+        Args:
+            tape (.JacobianTape): a quantum tape to apply the JAX interface to
+
+        **Example**
+
+        >>> with JacobianTape() as tape:
+        ...     qml.RX(0.5, wires=0)
+        ...     expval(qml.PauliZ(0))
+        >>> JAXInterface.apply(tape)
+        >>> tape
+        <JAXQuantumTape: wires=<Wires = [0]>, params=1>
+        """
+        tape_class = getattr(tape, "__bare__", tape.__class__)
+        tape.__bare__ = tape_class
+        tape.__class__ = type("JAXQuantumTape", (cls, tape_class), {})
+        return tape

```

Comment:
In retrospect, this is a design decision I do not like. Even ignoring the dynamic class change, this feels more like a function than a class method?

Reply:
I'm usually always against inplace transformations, and this one is pretty intense. tbh I don't really understand how this works.

Reply:
iirc the in-place transform was solely to placate Autograd. Everything else worked at the time without it being in-place (tf, torch), but autograd would give a really bizarre and long traceback that vanished when it was made in-place instead.

Reply:
(it could have been torch that complained rather than autograd, I can't remember fully).